flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/result/MaterializedCollectStreamResult.java

```diff
@@ -112,30 +129,43 @@ else if (!isRetrieving()) {
 	@Override
 	protected void processRecord(Tuple2<Boolean, Row> change) {
 		synchronized (resultLock) {
-			final Row row = change.f1;
 			// insert
 			if (change.f0) {
-				materializedTable.add(row);
-				rowPositionCache.put(row, materializedTable.size() - 1);
+				processInsert(change.f1);
 			}
 			// delete
 			else {
-				// delete the newest record first to minimize per-page changes
-				final Integer cachedPos = rowPositionCache.get(row);
-				final int startSearchPos;
-				if (cachedPos != null) {
-					startSearchPos = Math.min(cachedPos, materializedTable.size() - 1);
-				} else {
-					startSearchPos = materializedTable.size() - 1;
-				}
-
-				for (int i = startSearchPos; i >= 0; i--) {
-					if (materializedTable.get(i).equals(row)) {
-						materializedTable.remove(i);
-						rowPositionCache.remove(row);
-						break;
-					}
-				}
+				processDelete(change.f1);
+			}
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private void processInsert(Row row) {
+		// limit the materialized table
+		if (materializedTable.size() >= maxRowCount) {
+			processDelete(materializedTable.get(0));
```

Comment:
Why should we search for the row if we know that we want to remove it from the first position?
Couldn't we just do `materializedTable.remove(0)`? This will be expensive because `materializedTable` is an `ArrayList` but what's the chance that there is another row that is equal to the one that is removed?

Reply:
Maybe we should use a `LinkedList` instead?

Reply:
Or we just set elements to `null` and have an `startOffset` counter that we increase.
Once the counter equals `1000` (magic number!), we shrink the `ArrayList`.